runtime/src/main/java/org/corfudb/runtime/view/QuorumFutureFactory.java

```diff
@@ -0,0 +1,139 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.corfudb.runtime.view;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.concurrent.*;
+
+/**
+ * Factory for custom futures used by the quorum replication.
+ * Created by Konstantin Spirov on 2/3/2017.
+ */
+@Slf4j
+class QuorumFutureFactory {
+
+    /**
+     * Get a future that will complete only when n/2+1 futures complete or n/2+1 futures are canceled.
+     *
+     * The future returned does not block explicitly, it aggregates the futures and delegates the blocking.
+     *
+     * In case of normal execution, any of the compete futures can be used to return the result.
+     * In case of termination, the cancel flag will be updated and if any of the futures threw an exception,
+     * ExecutionException will be thrown,  otherwise the future will return null
+     *
+     * @param futures The N futures
+     * @return The composite future
+     */
+    static <R> Future<R> getQuorumFuture(CompletableFuture<R>... futures) {
+        return getCompositeFuture(futures.length/2+1, futures);
+    }
+
+    /**
+     * Get a future that will complete only when a single futures complete or n/2+1 futures are canceled.
```

Comment:
Actually (as we discussed earlier in this thread) there are 3 cases:
1) Quorum is reached
2) Quorum is not reached due to different values. . We should return exception.
3) Quorum is not reached due to failure in more than one node.

In both cases 2 and 3 we should return exception - so I agree with your remark (even when canceling you should not return null, this is inconsistent). 

Also there was inconsistency from the fact you have different variables futuresToWait and quorum (calculated here). 

In the next version I am preparing, the code here is simply the following:

 boolean noMoreHope = numUncompletedFutures+greatestNumCompleteFutures < futuresToWait;
  if (noMoreHope) {
     done = canceled = true;
       ..... // throw exception

So we are avo